go/mysql/binlog_event_json.go

```diff
@@ -17,490 +17,611 @@ limitations under the License.
+	jsonLargeArray  = 3
+	jsonLiteral     = 4
+	jsonInt16       = 5
+	jsonUint16      = 6
+	jsonInt32       = 7
+	jsonUint32      = 8
+	jsonInt64       = 9
+	jsonUint64      = 10 //0x0a
+	jsonDouble      = 11 //0x0b
+	jsonString      = 12 //0x0c a utf8mb4 string
+	jsonOpaque      = 15 //0x0f "custom" data
+)
+
+func jsonDataTypeToString(typ uint) string {
+	switch typ {
+	case jsonSmallObject:
+		return "sObject"
+	case jsonLargeObject:
+		return "lObject"
+	case jsonSmallArray:
+		return "sArray"
+	case jsonLargeArray:
+		return "lArray"
+	case jsonLiteral:
+		return "literal"
+	case jsonInt16:
+		return "int16"
+	case jsonUint16:
+		return "uint16"
+	case jsonInt32:
+		return "int32"
+	case jsonUint32:
+		return "uint32"
+	case jsonInt64:
+		return "int64"
+	case jsonUint64:
+		return "uint64"
+	case jsonDouble:
+		return "double"
+	case jsonString:
+		return "string"
+	case jsonOpaque:
+		return "opaque"
+	default:
+		return "undefined"
 	}
-	result.WriteByte(')')
-	return nil
 }
 
-func printJSONArray(data []byte, large bool, result *bytes.Buffer) error {
-	pos := 0
-	elementCount, pos := readOffsetOrSize(data, pos, large)
-	size, pos := readOffsetOrSize(data, pos, large)
-	if size > len(data) {
-		return vterrors.Errorf(vtrpc.Code_INTERNAL, "not enough data for object, have %v bytes need %v", len(data), size)
+// literals in the binary json format can be one of three types: null, true, false
+type jsonDataLiteral byte
+
+// this is how mysql maps the three literals (null, true and false) in the binlog
+const (
+	jsonNullLiteral  = '\x00'
+	jsonTrueLiteral  = '\x01'
+	jsonFalseLiteral = '\x02'
+)
+
+// in objects and arrays some values are inlined, others have offsets into the raw data
+var inlineTypes = map[jsonDataType]bool{
+	jsonSmallObject: false,
+	jsonLargeObject: false,
+	jsonSmallArray:  false,
+	jsonLargeArray:  false,
+	jsonLiteral:     true,
+	jsonInt16:       true,
+	jsonUint16:      true,
+	jsonInt32:       false,
+	jsonUint32:      false,
+	jsonInt64:       false,
+	jsonUint64:      false,
+	jsonDouble:      false,
+	jsonString:      false,
+	jsonOpaque:      false,
+}
+
+//endregion
+
+//region util funcs
+
+// readInt returns either 32-bit or a 16-bit int from the passed buffer. Which one it is, depends on whether the document is "large" or not
+// JSON documents stored are considered "large" if the size of the stored json document is
+// more than 64K bytes. For a large document all types which have their inlineTypes entry as true
+// are inlined. Others only store the offset in the document
+// This int is either an offset into the raw data, count of elements or size of the represented data structure
+// (This design decision allows a fixed number of bytes to be used for representing object keys and array entries)
+func readInt(data []byte, pos int, large bool) (int, int) {
+	if large {
+		return int(data[pos]) +
+				int(data[pos+1])<<8 +
+				int(data[pos+2])<<16 +
+				int(data[pos+3])<<24,
+			pos + 4
 	}
```

Comment:
add test for large case.

Reply:
TestPlayerJSONDocs in vplayer_flaky_test.go tests "large" docs (which are > 64KB in binary)